domain/application/state/resource.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/collections/set"
+
+	coreresource "github.com/juju/juju/core/resource"
+	"github.com/juju/juju/domain/application"
+	"github.com/juju/juju/domain/application/charm"
+	apperrors "github.com/juju/juju/domain/application/errors"
+	"github.com/juju/juju/internal/errors"
+)
+
+// buildResourceToAdd creates resources to add based on provided app and charm
+// resources.
+// Returns a slice of resourceToAdd and an error if any issues occur during
+// creation. Validates app resources against charm resources, warning if
+// mismatches are found.
+// Returns an error satisfying [apperrors.InvalidResourceArgs] if there is
+// unexpected resource in argument that are not defined in charm.
+func (st *State) buildResourceToAdd(appUUID, charmUUID string,
+	appResources []application.AddApplicationResourceArg,
+	charmResources map[string]charm.Resource) (
+	[]resourceToAdd, error) {
+	resources := make([]resourceToAdd, 0, len(appResources))
+	appResourceSet := set.NewStrings()
+	charmResourceSet := set.NewStrings()
+	for _, r := range appResources {
+		appResourceSet.Add(r.Name)
+		uuid, err := coreresource.NewUUID()
+		if err != nil {
+			return nil, errors.Capture(err)
+		}
+		resources = append(resources, resourceToAdd{
+			UUID:            uuid.String(),
+			ApplicationUUID: appUUID,
+			CharmUUID:       charmUUID,
+			Name:            r.Name,
+			Revision:        r.Revision,
+			Origin:          r.Origin.String(),
+			State:           "available", //todo(gfouillet): use state.go once #18460 landed
+			CreatedAt:       st.clock.Now(),
```

Comment:
In my opinion the time should be done in the service layer. i.e. we should do `Now` there and pass it in to state. For one it makes things easier to test (you know exactly what should be in state) and two in my opinion this is technically business logic.

Reply:
* It is easy to test (you can define a clock in the test).
* It would be business logic if it would have some synchronisation between application creation time and other times. It is not, since application doesn't seem to have such fields.

From my perspective is less a feature than a way to fetch resources and audit things, in general.

However, I will make a change: it make no sense that resources inserted at the same time get a different time.